hastory-server/src/Data/Hastory/Server/Utils.hs

```diff
@@ -2,14 +2,19 @@
 
 module Data.Hastory.Server.Utils where
 
+import Control.Monad.Except
 import Crypto.Hash (Digest, SHA256(..), hashWith)
 import qualified Data.ByteString as B
-import Data.Hastory.Types (Entry(..), ServerEntry(..), SyncRequest(..))
+import Data.Hastory.Types
 import qualified Data.Text as T
 import qualified Data.Text.Encoding as T
 import Data.Time.Format (defaultTimeLocale, formatTime, iso8601DateFormat)
 import Path (fromAbsDir)
 
+ensureWith :: MonadError e m => e -> Maybe a -> m a
+ensureWith _ (Just a) = pure a
+ensureWith e Nothing = throwError e
```

Comment:
`throwError` and `MonadError` is also something you want to avoid, for the same reasons as `ExceptT`.

Reply:
Hi Syd - I am happy to monomorphize this to: `ensureWith :: SeverError -> Maybe a -> HastoryHandler a`, but I'd like to still be able to use `throwError`, given that that is the idomatic way of throwing an error in servant's `Handler` monad. (`HastoryHandler` is just `ReaderT ServerSettings Handler`). If we wanted to avoid `throwError`, I'd have to go see what monomorphic function that name resolves to for `Handler`, which is do-able but seems to me to be more work than it is worth. Let me know what you think. 

On a related note, I don't think I understand the problem with `MonadError`, `ExceptT`, and `throwError` / `catchError`. Is there an article somewhere that explains the drawbacks?

Reply:
aha! In that case you can have `ensureWith` use the `ThrowAll` type class but I'd still rather have it use `Handler` monomorphically.
I don't have an article but we can discuss it in the next call.

Reply:
I think I found a suitable solution that is both easier to read and also monomorphic.